doc/developer/design/20220511_replicated_compute_introspection.md

```diff
@@ -0,0 +1,360 @@
+* Distribute sinking and sourcing of introspection data between compute workers.
+
+See [Future Improvements](#future-improvements) for our plans for achieving
+these goals in the future.
+
+## Description
+
+The general idea is that replicas sink their introspection dataflows (which they
+already maintain) to persistent storage. The resulting storage collections are
+exposed to the user as per-replica *introspection tables* in the system catalog.
+When these tables are referenced in queries, query planning translates these
+references to persist source reads. As persistent storage is shared between all
+replicas, any replica is able to access all introspection sources and yield
+consistent introspection data back to the user.
+
+![Introspection data flow](static/introspection-data-flow.png)
+
+### COMPUTE
+
+#### Sinking Introspection Data
+
+Compute replicas already maintain introspection dataflows that feed arrangements
+that can be peeked by ADAPTER. These dataflows are extended to also sink their
+output to persistent storage.
+
+Eventually there will be a STORAGE client for interacting with storage
+collections, but it doesn't exist yet. Instead, replicas need to use the
+`persist` API directly. `persist` only abstracts a single storage shard that
+cannot be written to by multiple compute workers, so all introspection data from
+the same dataflow will be sinked through the same worker initially. We can still
+distribute the sinking of different introspection dataflows between different
+workers.
+
+The proposed design does not strictly require the existing arrangements, but we
+keep them in around for two reasons:
+
+  * Because persist compaction is not implemented yet, introspection storage
+    collections can grow without bounds. We therefore don't want to enable
+    sinking to persist for long-running Materialize deployments (like cloud). To
+    still be able to collect introspection data for such deployments, we need to
+    keep the existing mechanism of peeking introspection arrangements around
+    (even if it only works reliably for single-replica clusters).
+  * Once we have sinks working for replicated clusters, we'll want to think
+    about using them to sink the introspection arrangements instead of the
+    hand-rolled solution proposed in this design.
+
+To sink introspection data, replicas need the following storage connection
+information:
+
+  * Blob URI
+  * Consensus URI
+  * A shard ID for every log variant
+
+This connection information is passed to replicas through the
+`CreateInstance` compute command.
+
+#### Changes to the Compute API
+
+The `LoggingConfig` contained in `ComputeCommand::CreateInstance` is extended to
+also contain the storage connection information mentioned above (details subject
+to bike-shedding):
+
+```rust
+pub struct LoggingConfig {
+    pub granularity_ns: u128,
+    pub active_logs: HashMap<LogVariant, ActiveLog>,
+    pub log_logging: bool,
+    pub blob_uri: String,
```

Comment:
IIUC storage has started putting together some minimal wrappers around persist configuration and such, so we don't have to switch all of it once they finish fleshing out storage collections. it's worth syncing with them to see if there's something appropriate to use here

Reply:
It looks like we can use the `CollectionMetadata` struct here, add a `shard_id` (or `shard_ids`) field to it, and change `LoggingConfig` to:

```
pub struct LoggingConfig {
     pub granularity_ns: u128,
     pub active_logs: HashMap<LogVariant, ActiveLog>,
     pub log_logging: bool,
}

pub struct ActiveLog {
     pub index_id: GlobalId,
     pub collection: CollectionMetadata,
 }
```